### ISSUE TITLE
PHP8 (8.0 and 8.1) support

### ISSUE THREAD

**Issue by VVD** [usvn/usvn#73]
Probably this isn't full list, but lets start. :-]
```
diff -ur usvn-1.0.10.orig/src/app/controllers/GroupadminController.php usvn-1.0.10/src/app/controllers/GroupadminController.php
--- usvn-1.0.10.orig/src/app/controllers/GroupadminController.php
+++ usvn-1.0.10/src/app/controllers/GroupadminController.php
@@ -64,7 +64,7 @@
                }
                try {
                        $group->save();
-                       foreach ($_POST['users'] as $user) {
+                       foreach ((array)$_POST['users'] as $user) {
                                $group->addUser($user);
                        }
                        $this->_redirect("/admin/group/");
diff -ur usvn-1.0.10.orig/src/library/USVN/Crypt.php usvn-1.0.10/src/library/USVN/Crypt.php
--- usvn-1.0.10.orig/src/library/USVN/Crypt.php
+++ usvn-1.0.10/src/library/USVN/Crypt.php
@@ -77,7 +77,7 @@
                $text = $plainpasswd.'$apr1$'.$salt;
                $bin = pack("H32", md5($plainpasswd.$salt.$plainpasswd));
                for($i = $len; $i > 0; $i -= 16) { $text .= substr($bin, 0, min(16, $i)); }
-               for($i = $len; $i > 0; $i >>= 1) { $text .= ($i & 1) ? chr(0) : $plainpasswd{0}; }
+               for($i = $len; $i > 0; $i >>= 1) { $text .= ($i & 1) ? chr(0) : $plainpasswd[0]; }
                $bin = pack("H32", md5($text));
                for($i = 0; $i < 1000; $i++) {
                        $new = ($i & 1) ? $plainpasswd : $bin;
diff -ur usvn-1.0.10.orig/src/library/USVN/Db/Table/Row/Project.php usvn-1.0.10/src/library/USVN/Db/Table/Row/Project.php
--- usvn-1.0.10.orig/src/library/USVN/Db/Table/Row/Project.php
+++ usvn-1.0.10/src/library/USVN/Db/Table/Row/Project.php
@@ -65,7 +65,7 @@
                        $rights = $table_groupstofilesrights->fetchAll($g);
                        $file_rights = new USVN_Db_Table_FilesRights();
                        foreach ($rights as $right) {
-                               $file_right = $file_rights->find($right->files_rights_id)->current();
+                               $file_right = $file_rights->find((array)$right->files_rights_id)->current();
                                if ($file_right->projects_id == $this->id) {
                                        $right->delete();
                                }
@@ -178,9 +178,9 @@
                if (empty($name) || preg_match('/^\s+$/', $name)) {
                        throw new USVN_Exception(T_("The project's name is empty."));
                }
-               if (!get_magic_quotes_gpc()) {
+//             if (!get_magic_quotes_gpc()) {
                        $name = addslashes($name);
-               }
+//             }
                if (!preg_match('/^[0-9a-zA-Z_\-\/]+$/', $name)) {
                        throw new USVN_Exception(T_("The project's name is invalid. A project's name can only include alpha-numeric characters and '-' or '_'."));
                }
diff -ur usvn-1.0.10.orig/src/library/USVN/Db/Table/Row/User.php usvn-1.0.10/src/library/USVN/Db/Table/Row/User.php
--- usvn-1.0.10.orig/src/library/USVN/Db/Table/Row/User.php          
+++ usvn-1.0.10/src/library/USVN/Db/Table/Row/User.php      
@@ -28,7 +28,7 @@
        {
                if (is_numeric($group)) {
                        $groups = new USVN_Db_Table_Groups();
-                       $group = $groups->find($group)->current();
+                       $group = $groups->find((array)$group)->current();
                }
                $group->addUser($this);
        }
diff -ur usvn-1.0.10.orig/src/library/USVN/FilesAccessRights.php usvn-1.0.10/src/library/USVN/FilesAccessRights.php
--- usvn-1.0.10.orig/src/library/USVN/FilesAccessRights.php
+++ usvn-1.0.10/src/library/USVN/FilesAccessRights.php
@@ -38,7 +38,7 @@
        public function findByPath($group_id, $path)
        {
         $path = str_replace('//', '/', $path);
-               if (strlen($path) == 0 || $path{0} !== '/') {
+               if (strlen($path) == 0 || $path[0] !== '/') {
                        throw new USVN_Exception(T_("Invalid path %s."), $path);
                }
                $response = array('read' => false, 'write' => false);
@@ -79,7 +79,7 @@
        public function setRightByPath($group_id, $path, $read, $write, $recursive = false)
        {
                $path = preg_replace('#[/]{2,}#', '/', $path);
-               if (strlen($path) == 0 || $path{0} !== '/') {
+               if (strlen($path) == 0 || $path[0] !== '/') {
                        throw new USVN_Exception(T_("Invalid path %s."), $path);
                }
                $table_files = new USVN_Db_Table_FilesRights();
diff -ur usvn-1.0.10.orig/src/library/Zend/Cache/Backend.php usvn-1.0.10/src/library/Zend/Cache/Backend.php
--- usvn-1.0.10.orig/src/library/Zend/Cache/Backend.php
+++ usvn-1.0.10/src/library/Zend/Cache/Backend.php
@@ -63,7 +63,7 @@
      */
     public function __construct(array $options = array())
     {
-        while (list($name, $value) = each($options)) {
+        foreach ($options as $name => $value) {
             $this->setOption($name, $value);
         }
     }
@@ -78,7 +78,7 @@
     public function setDirectives($directives)
     {
         if (!is_array($directives)) Zend_Cache::throwException('Directives parameter must be an array');
-        while (list($name, $value) = each($directives)) {
+        foreach ($directives as $name => $value) {
             if (!is_string($name)) {
                 Zend_Cache::throwException("Incorrect option name : $name");
             }
diff -ur usvn-1.0.10.orig/src/library/Zend/Cache/Core.php usvn-1.0.10/src/library/Zend/Cache/Core.php
--- usvn-1.0.10.orig/src/library/Zend/Cache/Core.php
+++ usvn-1.0.10/src/library/Zend/Cache/Core.php
@@ -143,7 +143,7 @@
             Zend_Cache::throwException("Options passed were not an array"
             . " or Zend_Config instance.");
         }
-        while (list($name, $value) = each($options)) {
+        foreach ($options as $name => $value) {
             $this->setOption($name, $value);
         }
         $this->_loggerSanity();
@@ -158,7 +158,7 @@
     public function setConfig(Zend_Config $config)
     {
         $options = $config->toArray();
-        while (list($name, $value) = each($options)) {
+        foreach ($options as $name => $value) {
             $this->setOption($name, $value);
         }
         return $this;
diff -ur usvn-1.0.10.orig/src/library/Zend/Ldap/Converter.php usvn-1.0.10/src/library/Zend/Ldap/Converter.php
--- usvn-1.0.10.orig/src/library/Zend/Ldap/Converter.php
+++ usvn-1.0.10/src/library/Zend/Ldap/Converter.php
@@ -69,7 +69,13 @@
      */
     public static function hex32ToAsc($string)
     {
-        $string = preg_replace("/\\\([0-9A-Fa-f]{2})/e", "''.chr(hexdec('\\1')).''", $string);
+        $string = preg_replace_callback(
+            "/\\\[0-9A-Fa-f]{2}/",
+            function ($matches) {
+                return chr(hexdec($matches[0]));
+            },
+            $string
+        );
         return $string;
     }
 
@@ -393,4 +399,4 @@
         }
         return $v;
     }
-}
\ No newline at end of file
+}
diff -ur usvn-1.0.10.orig/src/library/Zend/Loader/PluginLoader.php usvn-1.0.10/src/library/Zend/Loader/PluginLoader.php
--- usvn-1.0.10.orig/src/library/Zend/Loader/PluginLoader.php
+++ usvn-1.0.10/src/library/Zend/Loader/PluginLoader.php
@@ -128,7 +128,7 @@
         }
 
         $last = strlen($prefix) - 1;
-        if ($prefix{$last} == '\\') {
+        if ($prefix[$last] == '\\') {
             return $prefix;
         }
 
diff -ur usvn-1.0.10.orig/src/library/Zend/Registry.php usvn-1.0.10/src/library/Zend/Registry.php
--- usvn-1.0.10.orig/src/library/Zend/Registry.php
+++ usvn-1.0.10/src/library/Zend/Registry.php
@@ -203,7 +203,7 @@
      */
     public function offsetExists($index)
     {
-        return array_key_exists($index, $this);
+        return array_key_exists($index, (array)$this);
     }
 
 }
diff -ur usvn-1.0.10.orig/src/library/Zend/Tool/Project/Provider/DbAdapter.php usvn-1.0.10/src/library/Zend/Tool/Project/Provider/DbAdapter.php
--- usvn-1.0.10.orig/src/library/Zend/Tool/Project/Provider/DbAdapter.php
+++ usvn-1.0.10/src/library/Zend/Tool/Project/Provider/DbAdapter.php
@@ -97,9 +97,9 @@
         parse_str($dsn, $dsnVars);
 
         // parse_str suffers when magic_quotes is enabled
-        if (get_magic_quotes_gpc()) {
+/*        if (get_magic_quotes_gpc()) {
             array_walk_recursive($dsnVars, array($this, '_cleanMagicQuotesInValues'));
-        }
+        }*/
         
         $dbConfigValues = array('resources' => array('db' => null));
                  
```

**Comment by VVD** [comment]
Fixed create project.

**Comment by VVD** [comment]
@georgeto, maybe you can commit patches above?

**Comment by georgeto** [comment]
> @georgeto, maybe you can commit patches above?

I don't have access rights to this repository,
Or did you mean to create a PR from it? Mhm, but why would you need me for that?

**Comment by VVD** [comment]
> I don't have access rights to this repository,

This one.

> Or did you mean to create a PR from it? Mhm, but why would you need me for that?

More forks for god of forks!
I hate rule - make fork even for 1-liner.
Fork only if you really plan to work on the project and not on 1 patch.
See how many forks different projects have and how many of them are dead - 99%.

**Comment by VVD** [comment]
Fix of `src/library/USVN/Db/Table/Row/User.php`: edit user's groups.

**Comment by VVD** [comment]
Fix of `src/library/USVN/Db/Table/Row/Project.php`: remove group from project.

**Comment by thewebbooth** [comment]
Hi VVD,
I have a fork I have been working on today.  I have the installer working, project, user and group creation working.  I haven't tested it in anger yet.

Do check out my fork, if it is useful to you,  it is here

https://github.com/thewebbooth/usvn

**Comment by thewebbooth** [comment]
I've been using usvn with PHP8.1 to make projects and it is working for me.  I've put in a pull request for my changes.  My fork is here  https://github.com/thewebbooth/usvn

**Comment by VVD** [comment]
@julien-duponchelle, hello!
Did you commited all fixes from this issue and 1.0.11 have their?
